### PR TITLE
Support for unified pan/tilt telemetry

### PR DESCRIPTION
**Pull request by kmdalal**
### Summary of Changes

- Implemented PLEXIL support for the topic `/owl_msgs/pan_tilt_position` as specified in Google doc.
   - Note that this topic is only used for OWLAT Sim.  OceanWATERS still reads pan and tilt from the `/joint_states` topic, because this is more efficient.  (The unified topic republishes this information and adds some lag).
- PLEXIL now provides both radian and degree versions of pan/tilt lookups.
    - Note that the PLEXIL interface is independent of the ROS interface and still reads pan and tilt individually, which is most useful.

### Unrelated Changes

- Added `plans/unified` as a home for PLEXIL plans that can run on either testbed.
- Discovered and worked around a pre-existing bug in Plexil plan selection: a plan that finishes very quickly will break the interface because of synchronization issues.  More info in the code comments.
- Replaced the individual subscriber members with a container.
 - Factored shared members in OwInterface and OwlatInterface into PlexilInterface.
 - Added `const` to many member functions that are.
 - Formatting: removed trailing whitespace in some files (done automatically by Emacs), shortened some lines.

### Test

Two tests are needed, one using OceanWATERS, and one using OWLAT Sim.  We assume you have OWLAT Sim installed.  First, a clean build that supports OWLAT is needed:

```
$ catkin clean
$ catkin build -DOWLAT=ON
```

Test 1: OceanWATERS
1. Start any simulation world.
2. Launch `ow_exec.launch` and run the plan `TestTelemetry`.  You should see something close to the printout below.  Note that we're reading the pan/tilt positions at startup.  Feel free to move the antenna and run the plan again.
3. To test the changes to the plan selection interface, run any other plan, and then run `TestTelemetry` again.
```
[ INFO]: PLEXIL: Pan degees: -0.000373310720624672
[ INFO]: PLEXIL: Pan radians: -6.51550120789324e-06
[ INFO]: PLEXIL: Tilt degees: -0.665347767210253
[ INFO]: PLEXIL: Tilt radians: -0.0116125092086978
```

Test 2: OWLAT Sim
1. Start `roscore`.
2. Start OWLAT Sim: `launch_owlat Scoop`
3. Launch `owlat_exec.launch`
4. Run the same plan, `TestTelemetry`.  Since nothing is publishing to the topic, you should see the Plexil adapter default values:

```
[ INFO]: PLEXIL: Pan degees: 0
[ INFO]: PLEXIL: Pan radians: 0
[ INFO]: PLEXIL: Tilt degees: 0
[ INFO]: PLEXIL: Tilt radians: 0
```
5. Now, start publishing to the topic manually:
```
$ rostopic pub -r 10 /owl_msgs/pan_tilt_position owl_msgs/PanTiltPosition '{ header: {seq: 10, stamp: {secs: 1, nsecs: 1}, frame_id: "3"}, value: [1.0, 2.0] }'
```
6. Run the `TestTelemetry` plan again, and you should see this:
```
[ INFO]: PLEXIL: Pan degees: 57.2957795130823
[ INFO]: PLEXIL: Pan radians: 1
[ INFO]: PLEXIL: Tilt degees: 114.591559026165
[ INFO]: PLEXIL: Tilt radians: 2
```

### CODE REVIEW

**Comment by AstroStucky** [comment]
Please remember to mention the compiler flag required to run OWLAT plans. I got the tests to work by compiling with the following, but I do not believe this flag is mentioned anywhere except in old PRs. 
```
catkin build -DOWLAT=ON
```
Come to think about it, is there any reason our build does not compile the owlat binary by default?

File: ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp [hunk 1]
```diff
@@ -84,40 +84,58 @@ void PlexilPlanSelection::waitForPlan(){
...
+   finishes execution very quickly.  Seeking a solution from the
+   PLEXIL team as of 12/16/22.
+
+    // Workaround for allPlansFinished() not working on first plan.
+    if (m_first_plan) m_first_plan = false;
+
     // Times out after 3 seconds or the plan is registered as running.
     int timeout = 0;
-    while(OwExecutive::instance()->getPlanState() && timeout < 30){
+    while(OwExecutive::instance()->allPlansFinished() && timeout < 30){
```
Comment (AstroStucky) [inline ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp:99]: There are a lot of literals here that require doing math to realize their origin, so I suggest defining the constants `float LOOP_RATE = 10` and `float TIMEOUT = 3.0`
```suggestion
    while(OwExecutive::instance()->allPlansFinished() && timeout < static_cast<int>(LOOP_RATE * TIMEOUT)){
```

File: ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp [hunk 2]
```diff
@@ -84,40 +84,58 @@ void PlexilPlanSelection::waitForPlan(){
...
       ros::spinOnce();
       rate.sleep();
       timeout+=1;
       if(timeout % 10 == 0){
-        ROS_ERROR("Plan not responding, timing out in %i seconds", (3 - timeout/10));
+        ROS_ERROR("Plan not responding, timing out in %i seconds",
+                  (3 - timeout/10));
```
Comment (AstroStucky) [inline ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp:105]:
```suggestion
                  (TIMEOUT - timeout/LOOP_RATE));
```

File: ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp [hunk 3]
```diff
@@ -84,40 +84,58 @@ void PlexilPlanSelection::waitForPlan(){
 void PlexilPlanSelection::runCurrentPlan(){
   std_msgs::String status;
   ros::Rate rate(10); // 10 Hz for overall rate we are spinning
```
Comment (AstroStucky) [inline ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp:86]:
```suggestion
  float LOOP_RATE = 10.0;
  float TIMEOUT = 3.0;
  ros::Rate rate(LOOP_RATE); // 10 Hz for overall rate we are spinning
```

File: ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp (hunk 1 above)
Comment (AstroStucky) [reply ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp:99]: I just realized this is all commented out. Feel free to ignore this suggestion if this code is slated for removal.

File: ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp [hunk 4]
```diff
@@ -84,40 +84,58 @@ void PlexilPlanSelection::waitForPlan(){
...
+    m_planSelectionStatusPublisher->publish(status);
+  }
+  else {
+    // Unable to run the plan for some reason.
+    status.data = "FAILED:" + plan_array[0];
+    m_planSelectionStatusPublisher->publish(status);
+  }
+
+  // Delete the plan we just ran from plan array.
   plan_array.erase(plan_array.begin());
```
Comment (AstroStucky) [inline ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp:139]: Suggestion for future refactoring, not this PR. `plan_array` should be `m_plan_array`, and would probably make more sense as an `std::queue`.

**Review by AstroStucky — CHANGES_REQUESTED** [review]
The tests work. 

I found a bug where the wrong callback function is being subscribed. My other comments are optional or are refactor suggestions that do not need to happen in this PR.

File: ow_plexil/src/plexil-adapter/OwlatInterface.cpp [hunk 5]
```diff
@@ -110,45 +110,60 @@ void OwlatInterface::initialize()
...
+      (make_unique<ros::Subscriber>
+       (m_genericNodeHandle ->
+        subscribe("/owlat_sim/ARM_JOINT_TORQUES", qsize,
+                  &OwlatInterface::armJointTorquesCallback, this)));
+
+    m_subscribers.push_back
+      (make_unique<ros::Subscriber>
+       (m_genericNodeHandle ->
+        subscribe("/owlat_sim/ARM_JOINT_VELOCITIES", qsize,
+                  &OwlatInterface::armJointAnglesCallback, this)));
```
[the remark this answers is not in the document]
Comment (kmdalal) [reply ow_plexil/src/plexil-adapter/OwlatInterface.cpp:135]: Good catch, done.

File: ow_plexil/src/plexil-adapter/OwlatInterface.cpp [hunk 6]
```diff
@@ -110,45 +110,60 @@ void OwlatInterface::initialize()
...
-
-   m_armToolSubscriber = make_unique<ros::Subscriber>
-        (m_genericNodeHandle ->
-         subscribe("/owlat_sim/ARM_TOOL", qsize,
-         &OwlatInterface::armToolCallback, this));
+    m_subscribers.push_back
+      (make_unique<ros::Subscriber>
+       (m_genericNodeHandle ->
+        subscribe("/owlat_sim/ARM_JOINT_ANGLES", qsize,
+                  &OwlatInterface::armJointAnglesCallback, this)));
```
[the remark this answers is not in the document]
Comment (kmdalal) [reply ow_plexil/src/plexil-adapter/OwlatInterface.cpp:117]: I agree, and I want to refactor this, but am not fluent in the best C++ idioms to do so.  Am adding a TODO for future work.

File: ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp (hunk 1 above)
Comment (kmdalal) [reply ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp:99]: Good suggestion.  Adding todo for future work, which I think is best done when the commented-out code is replaced with a better solution that I'm working on with the Plexil team.

Comment (kmdalal) [reply ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp:99]: Good suggestion, done.

File: ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp (hunk 2 above)
Comment (kmdalal) [reply ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp:105]: Done

File: ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp (hunk 3 above)
Comment (kmdalal) [reply ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp:86]: Done (approximately like this).

File: ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp (hunk 4 above)
Comment (kmdalal) [reply ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp:139]: Added `m_` for now.  Added TODO for considering queue.

**Review by AstroStucky — APPROVED** [review]
The changes look good. 

A retest does not seem necessary, and I am currently experiencing an issue with my machine that is preventing me from running ROS. So I will approve this without a retest, but if you feel I should, please re-request and I will retest when I can.

**Comment by kmdalal** [comment]
> The changes look good.
> 
> A retest does not seem necessary, and I am currently experiencing an issue with my machine that is preventing me from running ROS. So I will approve this without a retest, but if you feel I should, please re-request and I will retest when I can.

Thanks - I don't think a retest is needed, especially as more of the same kind of changes are coming, and I will probably keep adding to the `TestTelemetry` plan.

**Pull request merged**